File: Exercise4.py
class newNode():
    def __init__(self, data):
        self.key = data
        self.left = None
        self.right = None

def inorder(temp):
    if temp:
        inorder(temp.left)
        print(temp.key)
        inorder(temp.right)

def insert_at_root(temp,key):
    a = newNode(key)
    a.right = temp
    a.left = temp.left
    temp.left = None
    return a

def insert_to_empty(root,n):
    queue = []
    if root:
        queue.append(root)
        while(queue):
            a = queue.pop(0)
            if(a.left):
                queue.append(a.left)
            else:
                a.left = newNode(n)
                break
            if(a.right):
                queue.append(a.right)
            else:
                a.right = newNode(n)
                break
        return root
    else:
        return None

if __name__ == '__main__':
    root = newNode(10)
    root.left = newNode(11)
    root.left.left = newNode(7)
    root.right = newNode(9)
    root.right.left = newNode(15)
    root.right.right = newNode(8)

    print("Inorder traversal before insertion at root", end = " ")
    print("\n")
    inorder(root)

    print("Inorder traversal after insertion at empty space", end = " ")
    print("\n")
    inorder(insert_to_empty(root,100))

    print("Inorder traversal after insertion at root", end = " ")
    print("\n")
    inorder(insert_at_root(root, 12))

File: Exercise4.cs
using System;
using System.ComponentModel.Design.Serialization;
using System.Security.Cryptography.X509Certificates;

namespace Exercise4
{
    class BinaryTree
    {
        
       internal class Node
        {
           internal  int key;
            internal Node left, right;

             public Node(int key)
            {
                this.key = key;
                left = null;
                right = null;
            }
        }
        static Node root;
        static Node temp = root;

        static void inorder(Node temp)
        {
            if(temp==null)
            {
                return;
            }
            else
            {
                inorder(temp.left);
                Console.WriteLine(temp.key);
                inorder(temp.right);
            }
        }

        static void insert(Node temp,int key)
        {
            Node _node = new Node(key);

            Node pointerNode = root;

            Node pointerY = null;

            while(pointerNode !=null)
            {
                pointerY = pointerNode;
                if(key< pointerNode.key)
                {
                    pointerNode = pointerNode.left;
                }
                else
                {
                    pointerNode = pointerNode.right;
                }
            }

            if(pointerY == null)
            {
                pointerY = _node;
            }

            else if(key < pointerY.key)
            {
                pointerY.left = _node;
            }
            else
            {
                pointerY.right = _node;
            }
        }
        static void Main(string[] args)
        {
            root = new Node(10);
            root.left = new Node(11);
            root.left.left = new Node(7);
            root.right = new Node(9);
            root.right.left = new Node(15);
            root.right.right = new Node(8);

            Console.WriteLine("inorder traversal before insertion");
            inorder(root);

            int key = 12;
            insert(root, key);

            Console.WriteLine("inorder traversal after insertion");
            inorder(root);

            Console.WriteLine("Hello World!");
        }
    }
}
